# Import

In [66]:
import pandas as pd

# Some functions

In [150]:
def is_first_high(string):
    return 'A' <= string[0] <= 'Я' or string[0] == 'Ё'

In [151]:
def is_first_low(string):
    return 'а' <= string[0] <= 'я' or string[0] == 'ё'

# Work

In [2]:
df = pd.read_csv('train.csv')

In [114]:
df.head(5)

Word  Label
0  Аалтонен      1
1       Аар      0
2     Аарон      0
3     ААРОН      0
4    Аарона      0

In [115]:
df.tail(5)

Word  Label
101403    Ёлкин      1
101404    ёлкой      0
101405     ёлок      0
101406  ёлочкой      0
101407  ёмкость      0

### Посмотрим на метки слов, у которых первая буква строчная

In [73]:
df[df.Word.apply(lambda x: is_first_low(x))].Label.value_counts(normalize=True)

0    0.977279
1    0.022721
Name: Label, dtype: float64

* Вывод: 97.7% слов со строчной первой буквой принадлежат классу 0

Посмотрим на слова, начинающие со строчной буквы, которые "якобы" являются фамилиями:

In [121]:
df[(df.Word.apply(lambda x: is_first_low(x))) & (df.Label == 1)]

Word  Label
43       аббревиатура      1
446         автокаско      1
475     автоматизация      1
486         автоматов      1
658             агент      1
673         агентство      1
749     агрегирования      1
809         адаптация      1
1052             азот      1
1320              акт      1
1363        активисты      1
1391             акты      1
1523        алгоритмы      1
1638        алкоголик      1
1875            альфа      1
1987       американцы      1
2008           аминов      1
2057      амортизация      1
2085        амплитуда      1
2317         англофил      1
2729         антрацит      1
2934           апплет      1
3382          арсенал      1
3433      артиллерист      1
3681           аспект      1
3690          аспекты      1
3926            атомы      1
3999        аудитория      1
4111       африканцев      1
4282           бабник      1
...               ...    ...
98022        школьник      1
98027       школьники      1
98214           шорох      1
98408            штаб      1
98504           штаты      1
98632          штучка      1
98635          штучки      1
98685             шум      1
98742           шутка      1
98904           щитки      1
99221        экзамены      1
99244        экзотика      1
99293       экономика      1
99308     экономистом      1
99312        экономка      1
99446    эксплуатация      1
99762         элемент      1
99850       эллипсоид      1
100134     энтузиазма      1
100177       эпидемия      1
100190        эпизоды      1
100542           этап      1
100597            это      1
100655            эхо      1
100686              ю      1
100948        явление      1
100954        явления      1
101046          языки      1
101235         японец      1
101389           ящик      1

[1266 rows x 2 columns]

Что-то эти слова не похожи на фамилии!

* #### Ну давайте предположим, что все слова, начинающиеся со строчной буквы, принадлежат классу 0

###  Посмотрим на метки слов, у которых первая буква заглавная

In [154]:
df[(df.Word.apply(lambda x: is_first_high(x)))].Label.value_counts()

0    36316
1     9372
Name: Label, dtype: int64

In [82]:
df[(df.Word.apply(lambda x: is_first_high(x)))].Label.value_counts(normalize=True)

0    0.79487
1    0.20513
Name: Label, dtype: float64

Тут уже нет такого явного перевеса

* Посмотрим на слова из большего(нулевого) класса из этих слов

In [128]:
df[(df.Word.apply(lambda x: is_first_high(x))) & (df.Label == 0)]

Word  Label
1                  Аар      0
2                Аарон      0
3                ААРОН      0
4               Аарона      0
6               Аароне      0
7              Ааронов      0
8               Аахена      0
13             Абажуры      0
16             Абакана      0
31            Аббатиса      0
34           Аббатисой      0
39           Аббатстве      0
45              Абвилю      0
46             Абделем      0
48               Абдул      0
49              Абдуле      0
50             АБДУЛЛА      0
52             Абдулле      0
53             Абдуллу      0
55           Абдуловна      0
57             Абдулом      0
58         Абдурахмана      0
59        Абдурахманов      0
60      Абдурахманович      0
61        Абдурахманом      0
66               Абеля      0
67             АБЕРДИН      0
77            Абигайль      0
78             Абигейл      0
79            Абиджана      0
...                ...    ...
101320     Ясновидящим      0
101321     Ясновидящих      0
101322         ЯСНОСТИ      0
101323         Ясность      0
101325             Ясс      0
101326          Ястреб      0
101327         Ястреба      0
101333        Ястребов      0
101336         Ястребы      0
101340           Яунде      0
101341           Яффой      0
101343          Яхроме      0
101345             Яхт      0
101347            ЯХТЕ      0
101350         Яхтсмен      0
101354            ЯХТЫ      0
101358          ЯЧЕЙКИ      0
101360          Ячейку      0
101362          Ячмень      0
101363          ЯЧМЕНЬ      0
101365          Ячменя      0
101369            Яшке      0
101372            Яшма      0
101388            ЯЩИК      0
101390           Ящика      0
101391           ЯЩИКА      0
101392          ЯЩИКАМ      0
101394           ЯЩИКИ      0
101395          ЯЩИКОВ      0
101397          Ящиком      0

[36316 rows x 2 columns]

Ну вроде бы и вправду тут нет фамилий, но на счет этих еще не понятно: Абдурахманов, Абдурахманович

* Теперь посмотрим на слова из первого класса из этих слов

In [81]:
df[(df.Word.apply(lambda x: is_first_high(x))) & (df.Label == 1)]

Word  Label
0          Аалтонен      1
5            Аарона      1
9           Абабков      1
18         Абакумов      1
19         Абалкина      1
20       Абатуровым      1
21           Абашев      1
22         Абашидзе      1
23         Абашкина      1
24            Аббас      1
25            АББАС      1
26           Аббаса      1
27          Аббасов      1
41          Абботом      1
47       Абдрашитов      1
51      АБДУЛЛАЕВЫХ      1
54         Абдулова      1
56        Абдуловым      1
62           Абелев      1
63          Абелевы      1
64          АБЕЛЕВЫ      1
65         АБЕЛЕВЫХ      1
68         Абернети      1
81          Абидину      1
82           Абиева      1
93           Абнеру      1
131     Абрамовичей      1
133         Абрамсе      1
134        Абрамсов      1
135       Абрамсона      1
...             ...    ...
101111   Якобссоном      1
101115     Яковенко      1
101121     Яковлеву      1
101135       Якунин      1
101136       ЯКУНИН      1
101188          Янг      1
101189         Янга      1
101190        Янгам      1
101191      Янгблад      1
101192     Янгблада      1
101193         Янге      1
101194        Янгов      1
101195        ЯНГОМ      1
101203        Янеза      1
101204        ЯНЕЗА      1
101205         Янес      1
101211    Янковский      1
101212    Янковским      1
101213   Янковскому      1
101220      Янсеном      1
101221         Янси      1
101222       Янссен      1
101225     ЯНУКОВИЧ      1
101226    Януковича      1
101227    Януковичу      1
101284       ЯРОВАЯ      1
101324      Яспером      1
101367         Яшин      1
101368       Яшиным      1
101403        Ёлкин      1

[9372 rows x 2 columns]

Ну более менее похоже на правду

* видно, что корни слов во многих случаях совпадают

по-хорошему надо для похожих слов найти максимальный префикс, но что-то пока сложно

поэтому ограничимся первыми 5 символами

так же выберем те префиксы, относительны частоты которых в классе 1 больше чем в 0

In [199]:
clazz_1 = dict()
for word in df[(df.Word.apply(lambda x: is_first_high(x))) & (df.Label == 1)].Word.values:
    st = word[:5].lower()
    if st in clazz_1:
        clazz_1[st] += 1
    else:
        clazz_1[st] = 1

In [202]:
clazz_2 = dict()
for word in df[(df.Word.apply(lambda x: is_first_low(x))) | (df.Label == 0)].Word.values:
    st = word[:5].lower()
    if st in clazz_2:
        clazz_2[st] += 1
    else:
        clazz_2[st] = 1

In [206]:
start_words = set()
for key in clazz_1:
    if key in clazz_2:
        if clazz_1[key] / 9372 > clazz_2[key] / 36316:
            start_words.add(key)
    else:
        start_words.add(key)         

# Test

In [209]:
test_df = pd.read_csv('test.csv')

In [219]:
def predict(x):
    if is_first_low(x): return 0
    if x.lower()[:5] in start_words:
        return 1
    return 0

In [220]:
test_df['Prediction'] = test_df.Word.apply(lambda x: predict(x))

In [232]:
submission_df = pd.read_csv('sample_submission.csv')

In [233]:
submission_df.Prediction = test_df['Prediction']

In [231]:
submission_df.to_csv('answer.csv', index=False)